packages/embed-react/src/contexts/token-context.tsx

```diff
@@ -7,36 +7,114 @@ import {
+  createContextAndHook<
+    | {
+        _tag: "success";
+        tokenFetchState: TokenFetchState;
+        token?: WealthSweetToken;
+      }
+    | { _tag: "error"; tokenFetchState: TokenFetchState; error: TokenError }
+  >("TokenContext");
 
+/**
+ * Constant representing one minute in milliseconds.
+ */
 const ONE_MINUTE = 1000 * 60;
 
+/**
+ * Props for the TokenProvider component.
+ */
+export type TokenProviderProps = PropsWithChildren<{
+  /** Function to fetch a new token */
+  fetchToken: () => Promise<WealthSweetToken>;
+  /** Optional callback for handling token fetch errors */
+  onFetchTokenError?: (error: TokenError) => void;
+}>;
+
+/**
+ * Provider component for managing token state and fetching. Handles automatic token refresh
+ * and error handling for token generation.
+ * @param props - The component props
+ * @param props.children - Child components that will have access to the token context
+ * @param props.fetchToken - Function that returns a Promise resolving to a new WealthSweetToken
+ * @param props.onFetchTokenError - Optional callback that will be called if token fetching fails
+ */
 export function TokenProvider({
   children,
   fetchToken,
-}: PropsWithChildren<{ fetchToken: () => Promise<WealthSweetToken> }>) {
+  onFetchTokenError,
+}: TokenProviderProps) {
   const [token, setToken] = useState<WealthSweetToken>();
+  const [tokenFetchState, setTokenFetchState] =
+    useState<TokenFetchState>("INITIALISED");
+  const [error, setError] = useState<TokenError>();
+
+  /**
+   * Handles token fetch errors.
+   * @param error - The error that occurred during token fetching
+   */
+  const handleTokenError = useCallback(
+    (error: unknown) => {
+      const tokenErrror = {
+        message: "Failed to generate token",
+        error,
+      } as const;
+      setError(tokenErrror);
+      setTokenFetchState("ERROR");
+      if (onFetchTokenError) {
+        onFetchTokenError(tokenErrror);
+      }
+    },
+    [setError, onFetchTokenError, setTokenFetchState],
+  );
 
+  /**
+   * Generates a new token and schedules the next token generation.
+   */
   const generateToken = useCallback(async () => {
-    const token = await fetchToken();
-    setToken(token);
-    setTimeout(
-      () => {
-        generateToken().catch((err) => console.error(err));
-      },
-      token.expires - new Date().getTime() - ONE_MINUTE,
-    ); // One minute before this token expires, fetch a new token
-  }, [setToken, fetchToken]);
+    try {
+      setTokenFetchState("FETCHING");
+      const token = await fetchToken();
+      setToken(token);
+      setError(undefined);
+      setTokenFetchState("FETCHED");
+      // One minute before this token expires, fetch a new token
+      setTimeout(
+        generateToken,
+        token.expires - new Date().getTime() - ONE_MINUTE,
+      );
+    } catch (e) {
+      handleTokenError(e);
+    }
+  }, [setToken, fetchToken, setTokenFetchState]);
```

Comment:
I think its dangerous to have this side effect in the function. If anyone accidentally calls it twice or if the token is changed in any manner the state could become invalid.

I haven't had time to think of the cleanest solution but what about this?
```suggestion
    try {
      setTokenFetchState("FETCHING");
      const token = await fetchToken();
      setToken(token);
      setError(undefined);
      setTokenFetchState("FETCHED");
    } catch (e) {
      handleTokenError(e);
    }
  }, [setToken, fetchToken, setTokenFetchState]);

  // One minute before the current token expires, fetch a new token
  useEffect(() => {
    if (tokenExpiryTimer) {
      clearTimeout(tokenExpiryTimer);
    }
    if (token) {
      setTokenExpiryTimer(
        setTimeout(
          generateToken,
          token.expires - new Date().getTime() - ONE_MINUTE,
        ),
      );
    }
  }, [token]);
```

At the top add something like this

```
  const [tokenExpiryTimer, setTokenExpiryTimer] = useState<number>();
```